pennylane/devices/qubit/simulate.py

```diff
@@ -335,6 +360,359 @@ def simulate_partial(k):
+        mcm_samples[k] = mcm_samples[k][np.logical_not(mask)]
+
+
+def update_mcm_samples(op, samples, mcm_active, mcm_samples):
+    """Updates the mid-measurement sample dictionary given a MidMeasureMP and samples.
+
+    If the ``mcm_active`` dictionary is empty, we are at the root and ``mcm_samples`
+    is simply updated with ``samples``.
+
+    If the ``mcm_active`` dictionary is not empty, we need to find which samples are
+    corresponding to the current branch by looking at all parent nodes. ``mcm_samples`
+    is then updated with samples at indices corresponding to parent nodes.
+
+    To illustrate how the function works, let's take an example. Suppose there are
+    `2**20` shots in total and the computation is midway through the circuit at the
+    7th MCM, the active branch is `[0,1,1,0,0,1]` and each MCM everything happened to
+    split the counts 50/50 so there are `2**14` samples to update.
+    These samples are not contiguous in general and they are correlated with the parent
+    branches, so where do they go? They must update the `2**14` elements whose parent
+    sequence corresponds to `[0,1,1,0,0,1]`.
+    """
+    if mcm_active:
+        shape = next(iter(mcm_samples.values())).shape
+        mask = np.ones(shape, dtype=bool)
+        for k, v in mcm_active.items():
+            if k == op:
+                break
+            mask = np.logical_and(mask, mcm_samples[k] == v)
+        if op not in mcm_samples:
+            mcm_samples[op] = np.empty(shape, dtype=samples.dtype)
+        mcm_samples[op][mask] = samples
+    else:
+        mcm_samples[op] = samples
+
+
+def circuit_up_to_first_mcm(circuit):
+    """Returns two circuits; one that runs up-to the next mid-circuit measurement and one that runs beyond it.
+
+    Measurement processes are computed on each branch, and then combined at the node.
+    This can be done recursively until a single node is left.
+    This is true for `counts`, `expval`, `probs` and `sample` but not `var` measurements.
+    There is no way to recombine "partial variances" from two branches, so `var` measurements are replaced
+    by `sample` measurements from which the variance is calculated (once samples from all branches are available).
+
+    Args:
+        circuit (QuantumTape): The circuit to simulate
+
+    Returns:
+        QuantumTape: Circuit up to the first MCM and measuring the MCM samples if an MCM is found and ``circuit`` otherwise
+        (QuantumTape, None): Rest of the circuit
+        (MidMeasureMP, None): The first MCM encountered in the circuit
+    """
+    if not has_mid_circuit_measurements(circuit):
+        return circuit, None, None
+
+    # find next MidMeasureMP
+    def find_next_mcm(circuit):
+        for i, op in enumerate(circuit.operations):
+            if isinstance(op, MidMeasureMP):
+                return i, op
+        return len(circuit.operations) + 1, None  # pragma: no cover
+
+    i, op = find_next_mcm(circuit)
```

Comment:
Could we combine this with the `has_mid_circuit_measurements` check above? 

Basically,
```
if op is None:
    return circuit, None, None
```

? Might cut down on an O(n_ops) pass of the tape.

Reply:
Good idea.